### PR TITLE
Allow to specify a terraform provider version

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #347 #570 
| ❓ Documentation  | yes https://github.com/cloudskiff/driftctl-docs/pull/67

### CODE REVIEW

**Review by sjourdan — CHANGES_REQUESTED** [review]
Running driftctl by default without any option should download the latest provider release version. Should it be in this PR or in another one? 

```shell
~/.../cloudskiff/tf-aws ( master ?:4 ⊘ )
❯  rm -rf ~/.driftctl

~/.../cloudskiff/tf-aws ( master ?:4 ⊘ )
❯  driftctl-0.10.0 scan 
[...]
DEBU[0000] New provider library created
DEBU[0000] provider not found, downloading ... path=/home/sjourdan/.driftctl/plugins/linux_amd64/terraform-provider-aws_v3.19.0
Downloading terraform provider: aws
DEBU[0000] Downloading provider path=/home/sjourdan/.driftctl/plugins/linux_amd64 url=https://releases.hashicorp.com/terraform-provider-aws/3.19.0/terraform-provider-aws_3.19.0_linux_amd64.zip
DEBU[0004] Decompressing archive dst=/home/sjourdan/.driftctl/plugins/linux_amd64 src=/tmp/terraform-provider104562435
DEBU[0005] Download successful
[...]
```

**Comment by eliecharra** [comment]
> Running driftctl by default without any option should download the latest provider release version

I don't think it's a good idea to always use the latest version, in this PR we fix the version to the previously hardcoded ones. We should discuss about always use the latest version but this can cause uncontrolled behaviors imho

**Comment by sjourdan** [comment]
Would it take long to add some minimal checks for the content of `DCTL_TF_PROVIDER_VERSION`? like at least it looks like x.y.z? 

```shell
❯  LOG_LEVEL=debug DCTL_TF_PROVIDER_VERSION=\..\ AWS_PROFILE=cloudskiff AWS_REGION=eu-west-3 driftctl-0.10.0 scan
DEBU[0000] Bound environment variable to flag env=tf_provider_version flag=tf-provider-version value=.. 
[...]
DEBU[0000] provider not found, downloading ... path=/home/sjourdan/.driftctl/plugins/linux_amd64/terraform-provider-aws_v.. 
Downloading terraform provider: aws
DEBU[0000] Downloading provider path=/home/sjourdan/.driftctl/plugins/linux_amd64 url=https://releases.hashicorp.com/terraform-provider-aws/.. AWS_PROFILE=cloudskiff/terraform-provider-aws_.. AWS_PROFILE=cloudskiff_linux_amd64.zip
unsuccessful request to https://releases.hashicorp.com/terraform-provider-aws/.. AWS_PROFILE=cloudskiff/terraform-provider-aws_.. AWS_PROFILE=cloudskiff_linux_amd64.zip: 403 Forbidden
```

File: pkg/terraform/provider_installer_test.go [hunk 1]
```diff
@@ -147,3 +143,40 @@ func TestProviderInstallerInstallAlreadyExistButIsDirectory(t *testing.T) {
 	)
 
 }
+
+// Ensure that if a provider exist with a postfix (_x5) we properly detect it
```
Comment (wbeuil) [inline pkg/terraform/provider_installer_test.go:147]: `exists`

**Comment by eliecharra** [comment]
> Would it take long to add some minimal checks for the content of DCTL_TF_PROVIDER_VERSION? like at least it looks like x.y.z?

No that's very easy, I need to handle a better validation of the input and also add a 404 / 403 error handling if the version does not exist

**Comment by wbeuil** [comment]
> > Running driftctl by default without any option should download the latest provider release version
> 
> I don't think it's a good idea to always use the latest version, in this PR we fix the version to the previously hardcoded ones. We should discuss about always use the latest version but this can cause uncontrolled behaviors imho

I understand why you would not want to always download the latest version of the provider but we should speak/find a way to bump sometimes the provider version and not be stuck at a former version. For that we could use our acc tests to act as our decider. At the end when we will be ok with the way we bump the version, let's put it in the docs for our user to understand our reasoning.

File: pkg/cmd/scan.go [hunk 2]
```diff
@@ -341,3 +353,13 @@ func parseOutputFlag(out string) (*output.OutputConfig, error) {
 		Options: options,
 	}, nil
 }
+
+func validateTfProviderVersionString(version string) error {
+	if version == "" {
+		return nil
+	}
+	if match, _ := regexp.MatchString("^\\d+\\.\\d+\\.\\d+$", version); !match {
+		return errors.Errorf("Invalid version argument %s, expected a valid semver string (e.g: 2.13.4)", version)
```
Comment (wbeuil) [inline pkg/cmd/scan.go:362]: `e.g. 2.13.4` the convention is the dot after each letter and then the examples

**Review by wbeuil — CHANGES_REQUESTED** [review]
Tested it with terraform provider aws 2.20.0 3.19.0 3.20.0 3.30.0 3.44.0 3.45.0.

Seems good to me.

Wonder what to do if someone tries to use driftctl with resource in one version and tf-provider in another and we find drifts.

Example I have: 

```
Found changed resources:
    - driftctl_assume_role (aws_iam_role):
        + inline_policy: <nil> => [map[name:driftctl_policy policy:{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "",
      "Effect": "Allow",
      "Action": [
        "sqs:ListQueues",
        "sqs:ListQueueTags",
        "sqs:GetQueueAttributes",
        "sns:ListTopics",
        ...
      ],
      "Resource": "*"
    }
  ]
}]]
Found 2 resource(s)
 - 100% coverage
 - 2 covered by IaC
 - 0 not covered by IaC
 - 0 missing on cloud provider
 - 1/2 changed outside of IaC
```

Only for tf-provider >= 3.30.0

**Comment by eliecharra** [comment]
> Wonder what to do if someone tries to use driftctl with resource in one version and tf-provider in another and we find drifts.

I don't know if we can do something for theses cases.

**Review by wbeuil — APPROVED** [review]

**Comment by sjourdan** [comment]
LGTM! 

## Invalid version argument

- `DCTL_TF_PROVIDER_VERSION=a` 
- `DCTL_TF_PROVIDER_VERSION=1..1.1`
- `DCTL_TF_PROVIDER_VERSION=a.b.c`
- `DCTL_TF_PROVIDER_VERSION=1`
- `DCTL_TF_PROVIDER_VERSION=٣`
- `DCTL_TF_PROVIDER_VERSION=٣.٣.٣`
- `DCTL_TF_PROVIDER_VERSION=\..\`

## Provider does not exist: 

- `DCTL_TF_PROVIDER_VERSION=199999999999999999999999999999999999999999999999999999.29999.999999`
- `DCTL_TF_PROVIDER_VERSION=latest`
- `DCTL_TF_PROVIDER_VERSION=666`

## Works

- `VERSION=3.0.0 DCTL_TF_PROVIDER_VERSION=${VERSION}`

**Review by sjourdan — APPROVED** [review]
LGTM

File: pkg/cmd/scan.go [hunk 3]
```diff
@@ -341,3 +353,13 @@ func parseOutputFlag(out string) (*output.OutputConfig, error) {
...
 	}, nil
 }
+
+func validateTfProviderVersionString(version string) error {
+	if version == "" {
+		return nil
+	}
+	if match, _ := regexp.MatchString("^\\d+\\.\\d+\\.\\d+$", version); !match {
+		return errors.Errorf("Invalid version argument %s, expected a valid semver string (e.g. 2.13.4)", version)
+	}
```
Comment (sundowndev) [inline pkg/cmd/scan.go:363]: Can't you use go-version to check the version here? https://github.com/cloudskiff/driftctl/blob/main/pkg/iac/terraform/state/versions.go#L6

**Review by sundowndev — CHANGES_REQUESTED** [review]

File: pkg/cmd/scan.go (hunk 3 above)
Comment (eliecharra) [reply pkg/cmd/scan.go:363]: I was wondering if we can use this lib but I don't think validation match what we really want there https://github.com/hashicorp/go-version/blob/master/version_test.go#L49

The simple regex was more relevant IMHO for this usage.

**Review by sundowndev — APPROVED** [review]

**Pull request merged**